src/typechecker.erl

```diff
@@ -570,25 +570,38 @@ glb_ty({type, _, record, _}, {type, _, record, _}, _A, _TEnv) ->
 %% For now going with a very crude approximation.
 glb_ty(Ty1 = {type, _, map, Assocs1}, Ty2 = {type, _, map, Assocs2}, A, TEnv) ->
     case {Assocs1, Assocs2} of
-        {any, _} -> ret(Ty2);
-        {_, any} -> ret(Ty1);
+        %% TODO: add a test case
+        {[?any_assoc], _} -> ret(Ty2);
+        {_, [?any_assoc]} -> ret(Ty1);
         _ ->
             %% TODO: Too simplistic!
             %% Map keys are used as is, not merged together even if it is possible.
+
+            %% These are only used to make the whole assoc easier to look up by key.
             MAssocs1 = maps:from_list([ {Key, As} || ?type(_, [Key, _]) = As <- Assocs1 ]),
             MAssocs2 = maps:from_list([ {Key, As} || ?type(_, [Key, _]) = As <- Assocs2 ]),
+
             %% TODO: repeated, extract to external fun
             IsMandatory = fun
                               (?type(map_field_exact, _)) -> true;
                               (_) -> false
                           end,
+
+            %% Mandatory associations (`map_field_exact') are key here because they define what is
+            %% required, but also what is forbidden in a map.
```

Comment:
Mandatory means required yes, but forbidden is every key which is not covered by any assoc (mandatory or not).

Good to have comments btw. :-)

Reply:
Yeah, what I wanted to highlight is that if a map has no `_ => _`, then mandatory assocs do not only define what has to be present in the map, but also what cannot go there. Maybe the wording is not 100% accurate.

From another perspective, if we go over optional assocs, it's not a big deal if we don't find a corresponding assoc in the other map. If we don't find one for mandatory assocs, we should return `type(none)` as the GLB. Please note, though, that we go over the list of mandatory assocs, but we then check if they're present in either mandatory or optional assocs in the other map.

Reply:
Let's say `#{integer() => integer()}`. It has no mandatory assoc, yet non-integers are forbidden.

    GLB(#{integer() => integer()}, #{1..5 => 1..5, foo => bar}) = #{1..5 => 1..5}

Reply:
Good example!

With the latest changes it's solved properly.
We're still prone to the order of associations in map types, so I brought back the sorting hack for comparisons. This problem manifested in

```erlang
         ?glb( ?t(#{ a := integer(), _ => _ }), ?t(#{ b := float(), _ => _ }),
               ?t(#{ a := integer(), b := float(), _ => _ }) )
```

where the generated map, depending on attribute order, might be `#{ a := integer(), b := float(), _ => _ }` or `#{ b := float(), a := integer(), _ => _ }`.

I'm thinking that to solve it properly maps' associations on generation or normalisation ought to be sorted according to the order of the type hierarchy: first on keys, then values, then if still ambiguous, using the default Erlang total ordering. I think this would make sure map types are always represented uniformly. WDYT?